### PR TITLE
Deferred followup refactors

### PR DESCRIPTION
**Pull request by crisbeto**
A couple of refactors around `defer` that we had discussed before plus one bug fix I found along the way. Includes the following changes:

### refactor(compiler): store deferred triggers as a map
Stores the `deferred` block triggers as a map instead of an array, because triggers can't be duplicated and because having to search through an array will be inconvenient later on.

I've also added a `DeferredBlock.visitAll` method to deduplicate the logic from the various visitor implementations.

### refactor(compiler): add more deferred validations
Adds validations for the following invalid deferred block structures:
* Duplicated triggers.
* Multiple `minimum` parameters on `placeholder` and `loading` blocks.
* Multiple `after` parameters on `loading` blocks.

### refactor(compiler): correctly identify lazy directives and pipes used after a nested defer block
Fixes that if a directive/pipe is used after a nested `defer` block, we weren't tracking it as lazy anymore. This was due to the fact that we were resetting the `isInDeferBlock` to false every time instead of the previous value.

### CODE REVIEW

File: packages/compiler/test/render3/view/binding_spec.ts [hunk 1]
```diff
@@ -338,6 +357,25 @@ describe('t2 binding', () => {
         'DirA', 'DirB', 'DirD', 'DirPlaceholder', 'DirC', 'DirLoading', 'DirE', 'DirError', 'DirF'
       ]);
     });
+
+    it('should identify pipes used after a nested defer block as being lazy', () => {
```
Comment (eneajaho) [inline packages/compiler/test/render3/view/binding_spec.ts:361]:
```suggestion
    it('should identify directives used after a nested defer block as being lazy', () => {
```

File: packages/compiler/src/render3/r3_ast.ts [hunk 2]
```diff
@@ -186,17 +186,55 @@ export class DeferredBlockError implements Node {
...
 }
 
+export interface DeferredBlockTriggers {
+  when?: BoundDeferredTrigger;
+  idle?: IdleDeferredTrigger;
+  immediate?: ImmediateDeferredTrigger;
+  hover?: HoverDeferredTrigger;
+  timer?: TimerDeferredTrigger;
+  interaction?: InteractionDeferredTrigger;
+  viewport?: ViewportDeferredTrigger;
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/r3_ast.ts:196]: I think we should consider supporting multiple triggers for hover, interaction and viewport. For example:
```
{#defer on interaction(myButton); on interaction(anotherButton)} ... {/defer}
```
or for cases when we want to capture interaction with the placeholder block (i.e. `on interaction` with no arguments) and also some other elements on a page:
```
{#defer on interaction; on interaction(anotherButton)} ... {/defer}
```

File: packages/compiler/src/render3/r3_deferred_triggers.ts [hunk 3]
```diff
@@ -141,27 +138,36 @@ class OnTriggerParser {
...
-          this.triggers.push(createHoverTrigger(parameters, sourceSpan));
+          trackTrigger(
+              'hover', this.triggers, this.errors, createHoverTrigger(parameters, sourceSpan));
           break;
 
         case OnTriggerType.VIEWPORT:
-          this.triggers.push(createViewportTrigger(parameters, sourceSpan));
+          trackTrigger(
+              'viewport', this.triggers, this.errors,
+              createViewportTrigger(parameters, sourceSpan));
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/r3_deferred_triggers.ts:170]: Quick question: can we have a private function on this class instead?
```suggestion
          this.trackTrigger(
              'viewport', createViewportTrigger(parameters, sourceSpan));
```

Comment (crisbeto) [reply packages/compiler/src/render3/r3_deferred_triggers.ts]: Good idea. Done.

File: packages/compiler/src/render3/r3_ast.ts (hunk 2 above)
Comment (AndrewKushnir) [reply packages/compiler/src/render3/r3_ast.ts:196]: Discussed offline: we'll investigate how to support this scenario, but separately. This is not a blocker for this PR.

**Review by AndrewKushnir — APPROVED** [review]
👍

**Review by jessicajaniuk — APPROVED** [review]
As usual, this looks fantastic.

**Comment by crisbeto** [comment]
Caretaker note: this is missing a language service approval, but the changes there are just a minor refactor to account for the new data structure. Also `deferred` blocks aren't enabled for users yet so nobody should be affected by it.

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit b1f96096d3bb21a77a6705a521a5c8f8a8df3fc4.

**Pull request closed without merging**